### DISCUSSION TITLE
Serial Forwarding "Connection Refused"

### DISCUSSION THREAD

**Discussion by gruedisueli** [iRobotEducation/create3_docs#339] [Questions] [answered]
I've been following along with [this tutorial](http://edu.irobot.com/learning-library/intro-to-ros-2-serial-communication-with-create-3).

I also referred to the general documentation [here](http://iroboteducation.github.io/create3_docs/webserver/serial-config/).

When I try to connect to the Create using a python script I get an error message: "ConnectionRefusedError: [Errno 111] Connection refused"

I've restarted the Create application as directed in the instructions. 

I've also checked the ports on my own machine to make sure communication through 8883 works (it does).

I tried to also connect to the irobot using the command "telnet [IP Address] [PORT]", and got an error "telnet: Unable to connect to remote host: Connection refused"

I have a suspicion that something about the Create configuration is currently blocking my connection but I can't figure out what. Suggestions? Thanks!

**Comment by ateaford** [comment]
The links you posted are not working.

**Reply by gruedisueli** [reply]
The links work fine if you copy/paste them in your address bar. I just checked them. They don't work if you click them, for reasons I do not know, and it must be a Github bug in terms of how their website interprets these specific links.

**Please copy paste them in your address bar and you will go to the correct page.**

**Comment by shamlian** [comment, marked as answer]
Hi there,

Per the documentation site you cited, the robot needs to be _power cycled_ for the beta feature to be enabled after clicking "Save" the first time. You only mentioned restarting the application. That will not work. Please power cycle your robot (turn it off by holding the power button for ~10 seconds, then put it on the dock) after clicking "save" and try again. Please note this should only be necessary to enable the feature the first time after a factory reset. Also, please be sure the robot adapter board's BLE/USB switch is set to the USB position; you can confirm this with log messages showing the device when it is plugged into the robot's USB-C port.

Please let us know if you have any more trouble!

PS: I fixed your links; if you edit your post, you'll be able to see an example of how markdown links work.

**Reply by gruedisueli** [reply]
Hi, actually I have power cycled it and restarted the application multiple times and it still has this issue. Thanks for fixing the links :)

**Reply by shamlian** [reply]
No problem, I'm here to help! Can you please post logs from your robot, plugging and unplugging the serial device, so I can try to see what is wrong? Can you please also confirm what device you are plugging into the robot, and that the USB/BLE switch on the adapter board is set to USB? I'm sorry to say I have not watched the tutorial. I can try to find time for that next week.

**Reply by gruedisueli** [reply]
Sure thing! I'm a bit of a newb with the Create3. Can you explain how to get these logs?

Also, is there a way to SSH to the Create3?

**Reply by gruedisueli** [reply]
OK, I found the logs in the Create Application. I didn't know there was a "logs" tab conveniently located there. I also switched the adapter board to "USB".

Here's what I get in the logs when I disconnect/reconnect the microcontroller I'm using:

```
Feb 24 16:46:04 iRobot-8954E52E3BB945B3BBEFAEA453B8AB8F kern.info kernel: [  385.862264] usb 2-1: USB disconnect, device number 2
Feb 24 16:46:22 iRobot-8954E52E3BB945B3BBEFAEA453B8AB8F kern.info kernel: [  405.022664] usb 2-1: new full-speed USB device number 3 using ohci-platform
Feb 24 16:46:22 iRobot-8954E52E3BB945B3BBEFAEA453B8AB8F kern.info kernel: [  405.308902] usb 2-1: New USB device found, idVendor=2341, idProduct=8057
Feb 24 16:46:22 iRobot-8954E52E3BB945B3BBEFAEA453B8AB8F kern.info kernel: [  405.308925] usb 2-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Feb 24 16:46:22 iRobot-8954E52E3BB945B3BBEFAEA453B8AB8F kern.info kernel: [  405.308935] usb 2-1: Product: Arduino NANO 33 IoT
Feb 24 16:46:22 iRobot-8954E52E3BB945B3BBEFAEA453B8AB8F kern.info kernel: [  405.308945] usb 2-1: Manufacturer: Arduino LLC
Feb 24 16:46:22 iRobot-8954E52E3BB945B3BBEFAEA453B8AB8F kern.info kernel: [  405.308957] usb 2-1: SerialNumber: 4964B96B50593730372E3120FF081B2F
Feb 24 16:46:22 iRobot-8954E52E3BB945B3BBEFAEA453B8AB8F kern.info kernel: [  405.311338] cdc_acm 2-1:1.0: ttyACM0: USB ACM device
```

**Reply by shamlian** [reply]
Can you post the full logs, please, so that I can see if the serial daemon is starting up correctly? Just click the "Download logs" button on the logs page. Also, can you confirm what you've saved in your serial configuration, and give me the version of firmware your Create 3 robot is running?

**Reply by shamlian** [reply]
Also (or maybe, do this first) -- have you tried again, since switching the BLE/USB switch to USB? There will be no port forward if there's no device to forward. Now that you have enabled the USB port by toggling that switch, if you have properly saved serial configuration settings, the port forward should start.

**Reply by gruedisueli** [reply]
Sorry, I didn't realize it, but actually somehow the original settings in the "Serial Configuration" page had all gotten erased at some point this morning as I was debugging things. Perhaps after I switched the BLE/USB switch, I don't know. Anyway, I navigated to the "Serial Configuration" page a few mins ago, found everything empty, and just typed in:
TTY device to foward: ttyACM0
Baud rate: 115200
External port number: 8883

And rebooted, and now things work!

**Reply by shamlian** [reply]
Glad everything is working!
